File: app/db/models/photo.py
from sqlalchemy import Column, String, DateTime, Text, Integer, ForeignKey
from sqlalchemy.dialects.postgresql import UUID, JSON
from db.database import Base
from sqlalchemy.orm import relationship
from uuid import uuid4
from datetime import datetime, timezone, timedelta

class Photo(Base):
    """
    photos 테이블 모델
    """
    __tablename__ = 'photos'

    # 사진 id
    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid4)
    # 사진명 (원본 파일명)
    name = Column(String, nullable=True)
    # 사진 저장소 주소 (Azure Blob Storage URL)
    url = Column(Text)
    # 연도
    year = Column(Integer)
    # 계절
    season = Column(String)
    # 설명
    description = Column(Text, nullable = True)
    # 관계 사용자 id
    user_id = Column(UUID(as_uuid=True), ForeignKey('users.id'))
    # 관계 가족 id
    family_id = Column(UUID(as_uuid=True), ForeignKey('families.id'))
    # 업로드 일자
    uploaded_at = Column(DateTime, default=lambda: datetime.now(timezone(timedelta(hours=9))))
    # Family ↔ Photo 
    family = relationship("Family", back_populates="photos")
    # User ↔ Photo
    user = relationship("User", back_populates="photos")
    # Photo ↔ Conversation 
    conversations = relationship("Conversation", back_populates="photo", cascade="all, delete-orphan")
    

File: lib/models/photo.dart
class Photo {
  final String id;
  final String? name;
  final String url;
  final int year;
  final String season;
  final String? description;
  final String familyId;
  final String uploadedAt;
  final String? sasUrl;
  final Map<String, dynamic>? user;

  Photo({
    required this.id,
    this.name,
    required this.url,
    required this.year,
    required this.season,
    this.description,
    required this.familyId,
    required this.uploadedAt,
    this.sasUrl,
    this.user,
  });

  factory Photo.fromJson(Map<String, dynamic> json) {
    return Photo(
      id: json['id'],
      name: json['name'],
      url: json['url'],
      year: json['year'],
      season: json['season'],
      description: json['description'],
      familyId: json['family_id'],
      uploadedAt: json['uploaded_at'],
      sasUrl: json['sas_url'],
      user: json['user'],
    );
  }

  String get formattedUploadedAt {
    try {
      final date = DateTime.parse(uploadedAt);
      return '${date.year}년 ${date.month.toString().padLeft(2, '0')}월 ${date.day.toString().padLeft(2, '0')}일';
    } catch (e) {
      return uploadedAt; // 파싱 실패 시 원본 반환
    }
  }
}